iothub/device/src/Transport/Mqtt/MqttTransportHandler.cs

```diff
@@ -4,1310 +4,1097 @@
-                {
-                    // .NET Core 2.0 doesn't support proxy. Ignore this setting.
-                    if (Logging.IsEnabled)
-                        Logging.Error(this, $"{nameof(CreateWebSocketChannelFactory)} PlatformNotSupportedException thrown as .NET Core 2.0 doesn't support proxy");
-                }
-
-                if (settings.WebSocketKeepAlive.HasValue)
-                {
-                    websocket.Options.KeepAliveInterval = settings.WebSocketKeepAlive.Value;
-                    if (Logging.IsEnabled)
-                        Logging.Info(this, $"{nameof(CreateWebSocketChannelFactory)} Set websocket keep-alive to {settings.WebSocketKeepAlive}");
-                }
-
-                if (connectionCredentials.Certificate != null)
-                {
-                    websocket.Options.ClientCertificates.Add(connectionCredentials.Certificate);
+                    systemProperties[propertyName] = ConvertFromSystemProperties(property.Value);
                 }
+            }
 
-                using var cts = new CancellationTokenSource(TimeSpan.FromMinutes(1));
-                await websocket.ConnectAsync(websocketUri, cts.Token).ConfigureAwait(false);
-
-                var clientWebSocketChannel = new ClientWebSocketChannel(null, websocket);
-
-                clientWebSocketChannel
-                    .Option(ChannelOption.Allocator, UnpooledByteBufferAllocator.Default)
-                    .Option(ChannelOption.AutoRead, false)
-                    .Option(ChannelOption.RcvbufAllocator, new AdaptiveRecvByteBufAllocator())
-                    .Option(ChannelOption.MessageSizeEstimator, DefaultMessageSizeEstimator.Default)
-                    .Pipeline.AddLast(
-                        MqttEncoder.Instance,
-                        new MqttDecoder(false, MaxMessageSize),
-                        new LoggingHandler(LogLevel.DEBUG),
-                        _mqttIotHubAdapterFactory.Create(this, connectionCredentials, additionalClientInformation, settings));
-
-                await s_eventLoopGroup.Value.RegisterAsync(clientWebSocketChannel).ConfigureAwait(true);
+            Dictionary<string, string> mergedProperties = systemProperties.Concat(message.Properties)
+                .ToLookup(x => x.Key, x => x.Value)
+                .ToDictionary(x => x.Key, g => g.First());
 
-                return clientWebSocketChannel;
-            };
-        }
-
-        private void ScheduleCleanup(Func<Task> cleanupTask)
-        {
-            Func<Task> currentCleanupFunc = _cleanupFunc;
-            _cleanupFunc = async () =>
-            {
-                await cleanupTask().ConfigureAwait(true);
+            string properties = UrlEncodedDictionarySerializer.Serialize(mergedProperties);
 
-                if (currentCleanupFunc != null)
-                {
-                    await currentCleanupFunc().ConfigureAwait(true);
-                }
-            };
-        }
```

Comment:
This conditional is personally pretty hard to read. Quite subjective opinion though that others are free to disagree with.

Reply:
Just want to clarify I mean line `1034`